// office-ui-fabric-react/lib-amd/components/ProgressIndicator/ProgressIndicator.styles.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = function (props) {
        var className = props.className, theme = props.theme, indeterminate = props.indeterminate, smoothTransition = props.smoothTransition;
        var palette = theme.palette, semanticColors = theme.semanticColors;
        return ({
            root: [
                'ms-ProgressIndicator',
                {},
                className
            ],
            itemName: [
                'ms-ProgressIndicator-itemName',
                {}
            ],
            itemDescription: [
                'ms-ProgressIndicator-itemDescription',
                {}
            ],
            itemProgress: [
                'ms-ProgressIndicator-itemProgress',
                {}
            ],
            progressTrack: [
                'ms-ProgressIndicator-progressTrack',
                {}
            ],
            progressBar: [
                'ms-ProgressIndicator-progressBar',
                {},
            ],
        });
    };
});
