// office-ui-fabric-react/lib-amd/components/ScrollablePane/ScrollablePane.styles.js
define(["require", "exports", "../../Styling"], function (require, exports, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = function (props) {
        var className = props.className;
        var AboveAndBelowStyles = {
            position: 'absolute',
            pointerEvents: 'auto',
            width: '100%',
            zIndex: 1
        };
        return ({
            root: [
                'ms-ScrollablePane',
                {
                    overflowY: 'auto',
                    maxHeight: 'inherit',
                    height: 'inherit',
                    WebkitOverflowScrolling: 'touch'
                },
                className
            ],
            stickyAbove: [
                {
                    top: 0,
                    selectors: (_a = {},
                        _a[Styling_1.HighContrastSelector] = {
                            borderBottom: '1px solid WindowText'
                        },
                        _a)
                },
                AboveAndBelowStyles
            ],
            stickyBelow: [
                {
                    bottom: 0,
                    selectors: (_b = {},
                        _b[Styling_1.HighContrastSelector] = {
                            borderTop: '1px solid WindowText'
                        },
                        _b)
                },
                AboveAndBelowStyles
            ]
        });
        var _a, _b;
    };
});
